### PR TITLE
Fixing user events merging

### PR DESCRIPTION
**Pull request by manasnisar**
## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

File: code/go/0chain.net/smartcontract/dbs/event/user.go [hunk 1]
```diff
@@ -142,23 +142,58 @@ func (edb *EventDb) updateUserPayedFees(users []User) error {
...
 }
 
 func mergeUpdateUserReadPoolTotalEvents() *eventsMergerImpl[ReadPoolLock] {
-	return newEventsMerger[ReadPoolLock](TagLockReadPool, withUniqueEventOverwrite())
+	return newEventsMerger[ReadPoolLock](TagLockReadPool, withReadPoolMerged())
+}
+
+func withReadPoolMerged() eventMergeMiddleware {
+	return withEventMerge(func(a, b *User) (*User, error) {
+		a.ReadPoolTotal += b.ReadPoolTotal
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/dbs/event/user.go]: `TagLockReadPool` emit event with data of `ReadPoolLock`, not `User`. Please also check others, and write unit tests to confirm that the changes work as expected.

**Review by peterlimg — DISMISSED** [review]

File: code/go/0chain.net/smartcontract/dbs/event/user.go (hunk 1 above)
Comment (manasnisar) [reply code/go/0chain.net/smartcontract/dbs/event/user.go]: Fixed.

Comment (manasnisar) [reply code/go/0chain.net/smartcontract/dbs/event/user.go]: Added unit tests

File: code/go/0chain.net/smartcontract/stakepool/lock.go [hunk 2]
```diff
@@ -113,7 +113,7 @@ func (sp *StakePool) LockPool(
 		Reward:       currency.Coin(0),
 		Total:        i,
 	}
-	balances.EmitEvent(event.TypeStats, event.TagLockStakePool, newPoolId, lock)
+	balances.EmitEvent(event.TypeStats, event.TagLockStakePool, txn.ClientID, lock)
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/stakepool/lock.go]: if client have different stake pools, those events will be merged even for different pools, is it the behavior we want?

File: code/go/0chain.net/smartcontract/storagesc/writepool.go [hunk 3]
```diff
@@ -87,7 +87,7 @@ func (ssc *StorageSmartContract) writePoolLock(
 
 	allocation.WritePool, err = currency.AddCoin(allocation.WritePool, txn.Value)
 	i, _ := txn.Value.Int64()
-	balances.EmitEvent(event.TypeStats, event.TagLockWritePool, allocation.ID, event.WritePoolLock{
+	balances.EmitEvent(event.TypeStats, event.TagLockWritePool, txn.ClientID, event.WritePoolLock{
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/storagesc/writepool.go]: write pool changes should be aggregated per allocation.ID

File: code/go/0chain.net/smartcontract/storagesc/writepool.go [hunk 4]
```diff
@@ -140,7 +140,7 @@ func (ssc *StorageSmartContract) writePoolUnlock(
 	}
 	alloc.WritePool = 0
 	i, _ := alloc.WritePool.Int64()
-	balances.EmitEvent(event.TypeStats, event.TagUnlockWritePool, alloc.ID, event.WritePoolLock{
+	balances.EmitEvent(event.TypeStats, event.TagUnlockWritePool, txn.ClientID, event.WritePoolLock{
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/storagesc/writepool.go]: the same

File: code/go/0chain.net/smartcontract/storagesc/readpool.go [hunk 5]
```diff
@@ -262,7 +262,7 @@ func (ssc *StorageSmartContract) readPoolUnlock(txn *transaction.Transaction, _
 	i, _ := balance.Int64()
 	key := readPoolKey(ssc.ID, txn.ClientID)
 
-	balances.EmitEvent(event.TypeStats, event.TagUnlockReadPool, key, event.ReadPoolLock{
+	balances.EmitEvent(event.TypeStats, event.TagUnlockReadPool, txn.ClientID, event.ReadPoolLock{
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/storagesc/readpool.go]: read pools are per client, so it is pretty the same, why do you want to change?

**Review by dabasov — CHANGES_REQUESTED** [review]

**Review by dabasov — APPROVED** [review]
lgtm

**Pull request merged**